### ISSUE TITLE
Standardize JSON handling of containers and loop plates

### ISSUE THREAD

**Issue by pauldhein** [ml4ai/delphi#211]
Currently we have two different styles for handing inputs for `container`s and `loop_plate`s in our JSON.

Containers:
```json
{
  "name": "update_est",
  "type": "container",
  "input": [
    {
      "name": "rain",
      "domain": "real"
    },
    {
      "name": "total_rain",
      "domain": "real"
    },
    {
      "name": "yield_est",
      "domain": "real"
    }
  ],
}
```

Loop plate
```json
{
  "name": "crop_yield__loop_plate__day_0",
  "type": "loop_plate",
  "input": [
    "consistency",
    "max_rain",
    "absorption",
    "rain",
    "total_rain",
    "yield_est"
  ],
}
```

1. Please change the style of the input list for `loop_plate`s to match that of a `container`s
2. `rain` is not defined before the call to`crop_yield__loop_plate__day_0` so it should not be an input variable. Can we please make this change?

The two items are also handled differently when they are being called:

A call for a container:
```json
{
  "function": "update_est",
  "output": {},
  "input": [
    {
      "variable": "rain",
      "index": 0
    },
    {
      "variable": "total_rain",
      "index": -1
    },
    {
      "variable": "yield_est",
      "index": -1
    }
  ]
},
```

A call for a loop plate:
```json
{
  "name": "crop_yield__loop_plate__day_0",
  "inputs": [
    "consistency",
    "max_rain",
    "absorption",
    "rain",
    "total_rain",
    "yield_est"
  ],
  "output": {}
},
```
3. Please change the `inputs` argument for a `loop_plate` to be named `input` and change the input list to match the input list schema for a `container`
4. Once again please remove `rain` as an input

Variable indexing for inputs is currently being handled differently:

Input variable indexing in a container:
```json
"body": [
  {
    "name": "update_est__assign__total_rain_1",
    "output": {
      "variable": "total_rain",
      "index": 1
    },
    "input": [
      {
        "variable": "total_rain",
        "index": 0
      },
      {
        "variable": "rain",
        "index": 0
      }
    ]
  },
]
```

Input variable indexing in a loop plate
```json
"body": [
  {
    "name": "crop_yield__assign__rain_0",
    "output": {
      "variable": "rain",
      "index": 0
    },
    "input": [
      {
        "variable": "day",
        "index": -1
      },
      {
        "variable": "consistency",
        "index": -1
      },
      {
        "variable": "max_rain",
        "index": -1
      },
      {
        "variable": "absorption",
        "index": -1
      }
    ]
  },
]
```

5. I prefer the indexing style for `loop_plate`s as it allows us to identify actual input variables by using -1 as the index as opposed to zero (which is the index for newly created variables in this function). Please change the container schema to use -1 as the index for input variables
6. I would like the loop index variable to not be indexed with -1. I would prefer to use `None` (`null` in JSON) for this value, as this can be used for me to easily identify the indexing variable. Please make this change as well.

@partikh please let me know if you have any questions regarding this issue.

**Comment by pratikbhd** [comment]
@pauldhein I am having some confusion trying to get a picture of how you want the new indexing with -1 to look like. 
Does this mean that the indexing should now start with -1 for all variable declarations and continue incrementing from there?

**Comment by pauldhein** [comment]
Sorry for the confusion @pratikbhd. I want `-1` to be used for inputs to a container or loop plate. Regular indexing will still start from 0. This will allow me to easily tell if a variable is an input, or a new variable defined in the container/loop plate. The reason this is necessary is that it is possible to see the following:
```python
x = 5
y = 2

def foo(a, b):
    c = a + b
    a = 4
    d = a + b + c

foo(x, y)
```

In this case, the first use of `a` in `foo` is referring to the input argument and thus it's index would be -1. The second use is a new definition of a variable named `a` inside of `foo` and its index would be 0.

**Comment by pratikbhd** [comment]
@pauldhein Revisiting this again, in the example you gave above, will the index of `a` in the third use i.e. `d = a + b + c` again be `-1`?
And if a fourth statement `a = 7` were to appear after `d = a + b + c`, would the index of `a` in that assignment be `1`?

**Comment by pauldhein** [comment]
So in the function the first use of `a` should get `-1` and the second use should get `0`. The reason the first use gets `-1` is because it is using the input `a`. Once `a` is redefined in the function then we will use that version of `a` (whose index should start at `0`.

Let's make this concrete with an example from `crop_yield.f`:
```fortran
      SUBROUTINE UPDATE_EST(RAIN, TOTAL_RAIN, YIELD_EST)
        DOUBLE PRECISION RAIN, YIELD_EST, TOTAL_RAIN
        TOTAL_RAIN = TOTAL_RAIN + RAIN
        IF(TOTAL_RAIN .le. 40) THEN
            YIELD_EST = -(TOTAL_RAIN - 40) ** 2 / 16 + 100
        ELSE
            YIELD_EST = -TOTAL_RAIN + 140
        ENDIF
      END SUBROUTINE UPDATE_EST
```
The first use of `TOTAL_RAIN` should have `-1` as its index and all subsequent uses should have `0` as their index (due to the redefinition of `TOTAL_RAIN`).

The big change that I want is I want to do the same for loop plates. Let's consider the loop plate in `crop_yield.f`:
```fortran
      DO 20 DAY=1,31
        RAIN = (-(DAY - 16) ** 2 / CONSISTENCY + MAX_RAIN) * ABSORPTION
        CALL UPDATE_EST(RAIN, TOTAL_RAIN, YIELD_EST)
        PRINT *, "Day ", DAY, " Estimate: ", YIELD_EST
   20 ENDDO
```
`CONSISTENCY`, `MAX_RAIN`, `ABSORPTION`, `TOTAL_RAIN`,  and `YIELD_EST` should all have `-1` as their index variables (because they are defined outside of the loop plate. `RAIN` should have `0` as its index variable because it is defined in the loop plate.

To answer your second question: by _loop index variable_ I am referring to `DAY`, the variable that tracks the increment through the loop. This variable will not have indexing in the style of the others (during GrFN wiring) so I'd like its index value to be set to `None` (`null` in JSON), such that I can easily differentiate it from the other variables in a loop plate.

Is all of this clear or should I add more explanation?

**Comment by pauldhein** [comment]
@adarshp I think I may have screwed up earlier when reviewing your changes. Since the git diff didn't show the whole file I think I may have requested that you alter changes that you made that were correct. This is in regards to adding an `index` field in the input list of a loop plate. Was the input list that you added this field to the input list associated with the definition of the loop plate function or the call to the loop plate function present in the body of the `crop_yield` container function? If it is the latter then your change was correct, and should be added to the codebase.

**Comment by pratikbhd** [comment]
@pauldhein Thanks for the detailed explanation. This certainly helps make things more clear. In order to nail this down, I will add some statements to the above example:
```fortran
      SUBROUTINE UPDATE_EST(RAIN, TOTAL_RAIN, YIELD_EST)
        DOUBLE PRECISION RAIN, YIELD_EST, TOTAL_RAIN
        TOTAL_RAIN = TOTAL_RAIN + RAIN
        IF(TOTAL_RAIN .le. 40) THEN
            YIELD_EST = -(TOTAL_RAIN - 40) ** 2 / 16 + 100
        ELSE
            YIELD_EST = -TOTAL_RAIN + 140
        ENDIF
        TOTAL_RAIN = YIELD_EST / 10
      END SUBROUTINE UPDATE_EST
```
In the above example, we will follow the index increment approach such that the index of `TOTAL_RAIN` in `TOTAL_RAIN = YIELD_EST / 10` will now be `1`, is that correct?

**Comment by pauldhein** [comment]
@pratikbhd that is correct!

**Comment by pratikbhd** [comment]
@pauldhein And in the case for the loop plates, if the index of `RAIN` will be `0` inside the loop plate, how will that affect it's indexing outside the loop plate? For example, if I had a program like the following:
```fortran
      SOME_VARIABLE = RAIN + 4
      RAIN =  56
      DO 20 DAY=1,31
        RAIN = (-(DAY - 16) ** 2 / CONSISTENCY + MAX_RAIN) * ABSORPTION
        CALL UPDATE_EST(RAIN, TOTAL_RAIN, YIELD_EST)
        PRINT *, "Day ", DAY, " Estimate: ", YIELD_EST
   20 ENDDO
       SOME_VARIABLE = RAIN + 6
```
In this case, would the index of `RAIN` in `SOME_VARIABLE = RAIN + 4` be `-1` (considering that it was the first use of `RAIN`) and its index in `RAIN =  56` be `0`. Then, `RAIN`'s index would jump back to `0` inside the loop plate (or will it now be `1`?)
Now after the loop plate, will the index of `RAIN` in `SOME_VARIABLE = RAIN + 6` be `0` (following from the loop plate) or will it be `1`?
I hope I am making sense here. Please let me know if this needs more clarification.

**Comment by pauldhein** [comment]
1. (A) No, `-1` is only to be used for input variables from an enclosing scope. Since this code snippet does not appear to have an enclosing scope, the index of `RAIN` in `SOME_VARIABLE = RAIN + 4` should be 0 (meaning the index of `RAIN` in `RAIN =  56` will be 1)
 (B) Inside the loop plate, when `RAIN` is redefined by `RAIN = (-(DAY - 16) ** 2 / CONSISTENCY + MAX_RAIN) * ABSORPTION` it will have the index 0 (because this is the first definition in that scope). I will be adding scope information to differentiate these two.
2. Oh that is a really good and really hard question to answer. Above all else we must preserve the functionality of the source language (Fortran in this case). We will need to ask @skdebray about this, but I believe it is the case that the value from inside the for loop would be used for the evaluation of `SOME_VARIABLE = RAIN + 6`. In that case I think we are going to need to consider the loop plates as having "outputs" of all the variables that are alive at the end of the loop. For the case of a variable that is defined before the beginning of the loop, its output variable will have the index increased by 1 as compared to the variable before the loop, irregardless of its value in the loop. Thus `RAIN` after the loop would have an index of 2. This is just my idea for now though. We should hash this out in greater detail at todays meeting.

**Comment by adarshp** [comment]
> @adarshp I think I may have screwed up earlier when reviewing your changes. Since the git diff didn't show the whole file I think I may have requested that you alter changes that you made that were correct. This is in regards to adding an `index` field in the input list of a loop plate. Was the input list that you added this field to the input list associated with the definition of the loop plate function or the call to the loop plate function present in the body of the `crop_yield` container function? If it is the latter then your change was correct, and should be added to the codebase.

It was in fact the latter. However, looks like @pratikbhd is incorporating the change into his own branch, 211_issue (from #221 ), so it's all good.

**Comment by pratikbhd** [comment]
@pauldhein 

> 1. (A) No, `-1` is only to be used for input variables from an enclosing scope. Since this code snippet does not appear to have an enclosing scope, the index of `RAIN` in `SOME_VARIABLE = RAIN + 4` should be 0 (meaning the index of `RAIN` in `RAIN =  56` will be 1)
>    (B) Inside the loop plate, when `RAIN` is redefined by `RAIN = (-(DAY - 16) ** 2 / CONSISTENCY + MAX_RAIN) * ABSORPTION` it will have the index 0 (because this is the first definition in that scope). I will be adding scope information to differentiate these two.
> 2. Oh that is a really good and really hard question to answer. Above all else we must preserve the functionality of the source language (Fortran in this case). We will need to ask @skdebray about this, but I believe it is the case that the value from inside the for loop would be used for the evaluation of `SOME_VARIABLE = RAIN + 6`. In that case I think we are going to need to consider the loop plates as having "outputs" of all the variables that are alive at the end of the loop. For the case of a variable that is defined before the beginning of the loop, its output variable will have the index increased by 1 as compared to the variable before the loop, irregardless of its value in the loop. Thus `RAIN` after the loop would have an index of 2. This is just my idea for now though. We should hash this out in greater detail at todays meeting.

On 1. (A), when you say enclosing scopes, do you also mean functions, like in this example?
```fortran
x = 5
y = 2

def foo(a, b):
    c = a + b
    a = 4
    d = a + b + c

foo(x, y)
```

**Comment by pauldhein** [comment]
Yep! Both loop plates and containers count as enclosing scopes. We __can__ have inputs to either of them and inputs to either of them should be handled in the same manner.  (Note that the outermost container will not have any inputs and thus will not have any identifiers with index `-1`).

**Comment by pratikbhd** [comment]
@pauldhein We discussed assignments inside loop plates and containers but what would the indexing look like for function calls?
For example, in `crop_yield.py` we have the following:
```fortran
DO 20 DAY=1,31
        RAIN = (-(DAY - 16) ** 2 / CONSISTENCY + MAX_RAIN) * ABSORPTION
        CALL UPDATE_EST(RAIN, TOTAL_RAIN, YIELD_EST)
        PRINT *, "Day ", DAY, " Estimate: ", YIELD_EST
```
Here, both `CALL` and `PRINT` are function calls. I am guessing the index for `RAIN`, TOTAL_RAIN` and `YIELD_EST` would following the same approach, right? So in this case, the indices for the inputs to `UPDATE_EST` would be:
```
RAIN: 0
TOTAL_PAIN: -1
YIELD_EST: -1
```
Is this correct?

**Comment by adarshp** [comment]
This seems to have been implemented, so I'm closing the issue.

**Comment by pratikbhd** [comment]
@adarshp This has not yet been implemented and I'm working on this on branch `211_issue`. The indexing changes for loop plates and function scopes are currently working but I'm trying to implement this for `IF` and nested `IF` functions.

**Comment by adarshp** [comment]
Ok, reopening. Please close when resolved.

**Comment by adarshp** [comment]
Closed by #232
